### PR TITLE
[SYNPY-1749]Allow quote, apostrophe and ellipsis in store_row_async

### PR DESCRIPTION
**Pull request by danlu1**
# **Problem:**

A JSON serialization issue occurs when a DataFrame passed to store_row_async contains a list or dictionary with strings that include both double quotes and apostrophes.

# **Solution:**

1. Add default value `"escapechar": "\\"` to `to_csv_kwargs` in `store_rows_async`
2. Add `to_csv_kwargs` to `_stream_and_update_from_df` so it can take the passed `to_csv_kwargs` values for downstream data processing. 

# **Testing:**

Unit test and integration test have been added.

### CODE REVIEW

**Comment by danlu1** [comment]
@andrewelamb sorry I should have marked this a draft.

File: synapseclient/models/mixins/table_components.py [hunk 1]
```diff
@@ -163,16 +165,74 @@ def convert_dtypes_to_json_serializable(df):
...
+                                return {
+                                    k: _reformat_special_values(v)
+                                    for k, v in obj.items()
+                                }
+                            if isinstance(obj, list):
+                                return [_reformat_special_values(item) for item in obj]
+                            return obj
+
+                        cleaned_x = _reformat_special_values(x)
+                        # return json.dumps(cleaned_x, ensure_ascii=False)
```
Comment (linglp) [inline synapseclient/models/mixins/table_components.py]: Nit: this is old code that needs to be removed?

**Review by linglp — CHANGES_REQUESTED** [review]
I think overall it looks good. The tests can be consolidated a bit to test all the edge cases in fewer integration tests to improve performance, and the docstring can be updated to reflect the new state of the code since json.dumps() was removed. There's also some logic that can be simplified in the redundant checks where sample_values is created but never actually used. The function name could also be more descriptive of what it actually does now (sanitizing special values rather than just converting dtypes)

File: tests/integration/synapseclient/models/async/test_table_async.py [hunk 2]
```diff
@@ -977,6 +982,410 @@ async def test_store_rows_as_large_df_being_split_and_uploaded(
         # AND The spy should have been called in multiple batches
         assert spy_send_job.call_count == 1
 
+    async def test_store_rows_with_json_dict_columns_and_quotes(
```
[the remark this answers is not in the document]
Comment (linglp) [reply tests/integration/synapseclient/models/async/test_table_async.py]: Based on the [contribution guide](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/CONTRIBUTING.md#lessons-learned-for-integration-testing), I think you should use scope="function" for test isolation. However, creating multiple tables with scope="function" is expensive, so consolidating these integration tests might be a better approach

[the remark this answers is not in the document]
Comment (linglp) [reply tests/integration/synapseclient/models/async/test_table_async.py]: Thanks Dan. I saw that the integration tests are now consolidated. Let me mark the comment as "resolved"

**Review by BryanFauble — CHANGES_REQUESTED** [review]
This is looking great, once we get the last few items handled (and develop merged in), I can approve!

File: tests/unit/synapseclient/mixins/unit_test_table_components.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+import json
```
Comment (BryanFauble) [inline tests/unit/synapseclient/mixins/unit_test_table_components.py]: Nit: `json` and `is_float_dtype` are imported but not used in the new tests.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

**Review by BryanFauble — CHANGES_REQUESTED** [review]
Nice work on this fix -- the backslash-escaping approach for embedded quotes makes sense, and the Ellipsis/pd.NA handling is solid. I flagged one issue that I think needs to be addressed before merge, plus a few cleanup nits.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

**Review by linglp — CHANGES_REQUESTED** [review]
Hi @danlu1 ! Thanks for your hard work. I think I found two bugs after a few rounds of testing myself: 

Bug 1: Nested `np.nan` not handled
```
df = pd.DataFrame({"val": [[1.0, np.nan], [2.0, 3.0]]})
```
The nested `np.nan` would currently pass through to JSON serialization, which could cause issues since np.nan is not valid JSON. 

This happened because`convert_dtypes()` only converts top-level `np.nan` to `pd.NA` but nested `np.nan` would remain unchanged. And even though in your code, you have: 
```
if obj is pd.NA:
```
But this won't handle `np.nan`. The fix would just be: 
```
def _reformat_special_values(obj):
    if pd.isna(obj):      # Catches pd.NA, np.nan, and None
        return None
```

Bug 2: Top-level missing values not handled
```
def _serialize_json_value(x):
    if isinstance(x, (list, dict)):
        # pd.NA handling is only here, inside _reformat_special_values
        ...
    if x is ...:
        return "..."
    return x  # <-- Top-level pd.NA, np.nan, None just pass through
```
Top-level `pd.NA` (or `np.nan`) isn't converted to None. It only works now because `.replace({pd.NA: None})` is called later.

**Review by BryanFauble — CHANGES_REQUESTED** [review]
This is great progress on a real user-reported issue, nice work on the recursive handling of Ellipsis and pd.NA in nested structures! There's one item that needs to be fixed before we can merge.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py [hunk 4]
```diff
@@ -3168,7 +3216,9 @@ async def store_rows_async(
         insert_size_bytes: int = 900 * MB,
         csv_table_descriptor: Optional[CsvTableDescriptor] = None,
         read_csv_kwargs: Optional[Dict[str, Any]] = None,
-        to_csv_kwargs: Optional[Dict[str, Any]] = None,
+        to_csv_kwargs: Optional[Dict[str, Any]] = {
+            "escapechar": "\\",
+        },
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py]: Good catch — fixed in `4fda36a8`. I also made it a merge rather than a replace (`{"escapechar": "\\", **(to_csv_kwargs or {})}`) so callers who pass their own `to_csv_kwargs` don't silently lose the `escapechar` default. Docstring updated to reflect the merge behavior.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py [hunk 5]
```diff
@@ -163,16 +165,63 @@ def convert_dtypes_to_json_serializable(df):
...
     for col in df.columns:
-        df[col] = (
-            df[col].replace({pd.NA: None}).astype(object)
-        )  # this will convert the int64 and float64 columns to object columns
+        sample_values = df[col].dropna()
+        if len(sample_values):
+            df[col] = df[col].apply(_serialize_json_value)
+            # restore the original values of the column especially for the int64 and float64 columns since apply function changes the dtype
+            df[col] = df[col].convert_dtypes()
+            df[col] = df[col].replace({pd.NA: None}).astype(object)
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py]: Good catch — fixed in `fbc50ec1`. Removed the `dropna() + len()` guard so every column goes through `convert_dtypes().replace({pd.NA: None}).astype(object)`. All-NA columns now end up with `object` dtype and `None` values as expected.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py [hunk 6]
```diff
@@ -163,16 +165,63 @@ def convert_dtypes_to_json_serializable(df):
...
+        # Handle standalone ellipsis
+        if x is ...:
+            return "..."
+        return x
 
     for col in df.columns:
-        df[col] = (
-            df[col].replace({pd.NA: None}).astype(object)
-        )  # this will convert the int64 and float64 columns to object columns
+        sample_values = df[col].dropna()
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py]: Agreed, simplified in `fbc50ec1` — the guard is gone and `apply()` runs unconditionally.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py]: Agreed — pushed this in `fbc50ec1`. `_serialize_json_value` now catches top-level `pd.NA`/`np.nan`/`None` directly via `pd.isna(x)` instead of relying on the downstream `.replace({pd.NA: None})` sweep.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py [hunk 7]
```diff
@@ -163,16 +165,63 @@ def convert_dtypes_to_json_serializable(df):
...
     """
-    import pandas as pd
+
+    def _serialize_json_value(x):
+        if isinstance(x, (list, dict)):
+
+            def _reformat_special_values(obj):
+                if obj is ...:
+                    return "..."
+                if obj is pd.NA:
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py]: Good catch — you're right, `np.nan` was slipping through since `obj is pd.NA` only matches `pd.NA`. Fixed in `fbc50ec1` by swapping to `pd.isna(obj)` (inside `_reformat_special_values`, after the list/dict recursion checks) so `pd.NA`, `np.nan`, and `None` all become `None` in nested structures.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

File: synapseclient/models/mixins/table_components.py [hunk 8]
```diff
@@ -163,16 +165,63 @@ def convert_dtypes_to_json_serializable(df):
...
+                    "references": [...]
+                },
+                {
+                    "id": 4,
+                    "description": 'Description containing apostrophes sage\'s',
+                    "references": [...]
+                }
+
+            ],
         }).convert_dtypes()
```
[the remark this answers is not in the document]
Comment (BryanFauble) [reply synapseclient/models/mixins/table_components.py:200]: Leaving the outer `.convert_dtypes()` calls in place for now — they run in `store_rows_async` before `_infer_columns_from_data` for schema inference, which happens before `convert_dtypes_to_json_serializable` is called downstream in `_chunk_and_upload_df`. Removing them could affect schema detection in a way that feels out of scope for this PR. Happy to take it as a follow-up if you'd like.

_Note: This comment was drafted with AI assistance and reviewed by me for accuracy._

**Review by BryanFauble — APPROVED** [review]
The last round of changes with my final patches looked good to me.

@danlu1 is going to give this a final review as well before we merge in.

**Review by danlu1 — COMMENTED** [review]
lgtm

File: synapseclient/models/mixins/table_components.py [hunk 9]
```diff
@@ -3560,6 +3620,8 @@ async def main():
         test_import_pandas()
         from pandas import DataFrame
 
+        to_csv_kwargs = {"escapechar": DEFAULT_ESCAPSE_CHAR, **(to_csv_kwargs or {})}
```
Comment (linglp) [inline synapseclient/models/mixins/table_components.py]: this is a typo right? it should be `DEFAULT_ESCAPE_CHAR` not `DEFAULT_ESCAPSE_CHAR`

File: synapseclient/models/mixins/table_components.py [hunk 10]
```diff
@@ -163,16 +174,64 @@ def convert_dtypes_to_json_serializable(df):
...
-    import pandas as pd
+
+    def _serialize_json_value(x):
+        if isinstance(x, (list, dict)):
+
+            def _reformat_special_values(obj):
+                if obj is ...:
+                    return "..."
+                if isinstance(obj, dict):
+                    return {k: _reformat_special_values(v) for k, v in obj.items()}
```
Comment (linglp) [inline synapseclient/models/mixins/table_components.py:212]: Is it safe to assume that the keys in the dictionaries don't need cleanup?

**Review by linglp — APPROVED** [review]
LGTM! Just one minor thing to be fix. I am going to approve now!

File: synapseclient/models/mixins/table_components.py (hunk 10 above)
Comment (danlu1) [reply synapseclient/models/mixins/table_components.py:212]: This is out of scope of the ticket since we don't know the format requirement or caught any errors regarding the table header.

File: synapseclient/models/mixins/table_components.py (hunk 9 above)
Comment (danlu1) [reply synapseclient/models/mixins/table_components.py]: Good catch. This is borrowed from table.py. I will fix it
